### PR TITLE
Upgrade rust and rework

### PR DESCRIPTION
**Pull request by npmenard**
As mentioned in yesterday's meeting this a rework of micro-rdk to make it a bit simpler. I bumped the minimum rust version to 1.75 which will break CI so we will have to do some synchronizing there. 
I also added a UDPMuxer instead of the former IO scheme we were using for transporting packets to sub part of webRTC i think it's more streamlined and definitely require less memory which is good.
You'll notice that most of the patched crates are gone as the changes we needed have made it upstream. We may need to still patch async-io since their max allocated timer is 100 and ours was set to 20 so I am not sure memory wise what the impact will be.

### CODE REVIEW

File: micro-rdk/src/common/webrtc/ice.rs
Comment (npmenard) [reply micro-rdk/src/common/webrtc/ice.rs:205]: i was fixing  a bug but it will be gone before the merge

File: micro-rdk/src/common/webrtc/udp_mux.rs
Comment (npmenard) [reply micro-rdk/src/common/webrtc/udp_mux.rs:174]: we currently only receive STUN or DTLS in the socket but potentially as we expand we may get other packets type

File: micro-rdk/src/common/webrtc/udp_mux.rs
Comment (npmenard) [reply micro-rdk/src/common/webrtc/udp_mux.rs:252]: the read interest flag is set when the consumer has been taken. The goal behind it is to avoid stalling  the pipeline if for example we get STUN messages expecting  DTLS but there are no STUN reader running

File: micro-rdk/src/esp32/entry.rs
Comment (npmenard) [reply micro-rdk/src/esp32/entry.rs:46]: too heavy weight. It's a workaround for now we still have to 100% there is a race and that this is the only way to fix it

File: micro-rdk/src/common/webrtc/udp_mux.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/webrtc/udp_mux.rs:257]: I thought you could force enums to be usize type and have integer values (then you can use them as the indices as well). Something like this?
```rust
#[repr(usize)]
enum Direction {
    STUN,
    TURN
}

pub(crate) fn get_stun_mux(&self) -> Option<UdpMux> {
        let state = &mut self.mux.lock().unwrap()[Direction::STUN as usize];
        if !state.interest {
            state.interest = true;
            Some(UdpMux {
                muxer: self.clone(),
                direction: Direction::STUN,
                peer_addr: None,
            })
        } else {
            None
        }
}
```

File: micro-rdk/src/common/webrtc/udp_mux.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/webrtc/udp_mux.rs:252]: I still think something like "consumer_available" with the comment "if false, an existing consumer is expecting to read from this mux" would be clearer.

File: micro-rdk/src/common/webrtc/udp_mux.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/webrtc/udp_mux.rs:257]: or you could implement a function on the enum like
```rust
const DTLS: usize = 0;
const STUN: usize = 1;

enum Direction {
    STUN,
    TURN
}

impl Direction {
    fn to_mux_index(&self) -> usize {
        match self {
            Self::STUN => STUN,
            Self::DTLS => DTLS
        }
    }
}
```
I don't think this makes it that much more complex

File: micro-rdk/src/common/webrtc/udp_mux.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/webrtc/udp_mux.rs:177]: ok, could we do a match statement with a case for `_ => unreachable!()` ? Feels less fragile

File: micro-rdk/src/common/webrtc/udp_mux.rs
Comment (npmenard) [reply micro-rdk/src/common/webrtc/udp_mux.rs]: agreed i renamed it to `is_listening`

File: micro-rdk/src/common/webrtc/udp_mux.rs
Comment (npmenard) [reply micro-rdk/src/common/webrtc/udp_mux.rs]: I implemented with this index trait. Bonus it actually generates less instructions in release mode than the original implementation !!

Comment (gvaradarajan) [reply micro-rdk/src/common/webrtc/udp_mux.rs]: yeah it looks cool

File: micro-rdk/src/esp32/hcsr04.rs
Comment (acmorrow) [inline micro-rdk/src/esp32/hcsr04.rs:89]: I think you can probably just discard the high bits of the result from `esp_timer_get_time` to get a 32 bit value and use that, as long as you add a check that `when` > `prior` below. In the rare case that your low bits roll over between the issuing and receiving the pulse, resulting in `when` < `prior`, just don't notify, and the consumer will time out.

File: etc/docker/Dockerfile
Comment (acmorrow) [inline etc/docker/Dockerfile:9]: Why isn't the old `binstall` mechanism appropriate anymore?

**Review by acmorrow — APPROVED** [review]
LGTM with a few small comments / suggestions / questions here and there.

File: micro-rdk/src/common/webrtc/udp_mux.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/webrtc/udp_mux.rs:205]: To clarify further, assuming that any message type 2 or greater is DTLS automatically feels like a code smell to me because it relies on an external assumption. Wouldn't it be better to raise a specific error here if for some reason we receive a message through the UDP socket of an unexpected packet type?

**Review by gvaradarajan — APPROVED** [review]
LGTM other than this one small thing

File: Cargo.toml
Comment (npmenard) [reply Cargo.toml:36]: I forgot I_think I had to pin it while i was upgrading other crates but removing it does break the build

File: Cargo.toml
Comment (npmenard) [reply Cargo.toml:60]: specifically for `futures-rustls` we have to pin it to 0.22 because later versions break the native build

Comment (npmenard) [reply Cargo.toml:60]: I will get to it at one point :P

File: micro-rdk/src/common/grpc.rs
Comment (npmenard) [reply micro-rdk/src/common/grpc.rs:1319]: !!

File: micro-rdk/src/common/webrtc/ice.rs
Comment (npmenard) [reply micro-rdk/src/common/webrtc/ice.rs:181]: No it's a shameless hack. Creating a BytesMut with capacity will not initialize it even if the memory is allocated. so the buffer will have len = 0. So we set the len to 256 to make it look like it was initialized

File: micro-rdk/src/common/webrtc/sctp.rs
Comment (npmenard) [reply micro-rdk/src/common/webrtc/sctp.rs:139]: The assumption holds because the dtls object is boxed it works because we know this but really is bad implementation.  Its actually a leftover when I upgraded and remove the Unpin constrain on S we don't need to do that so I am going the put it back

File: etc/docker/Dockerfile
Comment (npmenard) [reply etc/docker/Dockerfile:9]: it's annoying because it the maintainer of the crate we try do install don't properly set the metadata binstall will fail to find the right precompiled binary and revert de compiling from source. For amd64 it's okay for arm64 it makes building the image excessively long

File: micro-rdk/src/esp32/hcsr04.rs
Comment (npmenard) [reply micro-rdk/src/esp32/hcsr04.rs:89]: :ok_hand:

File: micro-rdk/src/common/webrtc/udp_mux.rs
Comment (npmenard) [reply micro-rdk/src/common/webrtc/udp_mux.rs:206]: I see you point but it's actually ok. If the packet is not a proper DTLS record it will be rejected by the DTLS layer. 
We shouldn't report an error anyway and rather drop the packet silently. I don't want to write a state machine or a filter that supports identify correct STUN or DTLS packet it feels a bit cumbersome in this scenario

File: micro-rdk/src/common/conn/server.rs
Comment (acmorrow) [inline micro-rdk/src/common/conn/server.rs:518]: I think you can take out the comment now.

File: micro-rdk/src/common/webrtc/candidates.rs
Comment (acmorrow) [inline micro-rdk/src/common/webrtc/candidates.rs:393]: Just double checking that this is intended, but if this branch is not taken you will continue to line 401 since there is no `else` case.

Comment (npmenard) [reply micro-rdk/src/common/webrtc/candidates.rs:393]: yup we create a new request if the preceding one was successful

**Pull request merged**
